### PR TITLE
Closes #2138 Implement derive_vars_extreme_event()

### LINKED ISSUE #2138: Feature Request: Implement `derive_vars_extreme_event()`
### Feature Idea

```
derive_vars_extreme_event(
  dataset,
  by_vars,
  order,
  mode,
  events, # expects a list of event() or event_joined() objects
  source_datasets,
  missing_values = NULL,
  tmp_event_nr_var = NULL, # the variable can be used in order
  check_type = "warning"
)
```
The function should work like `derive_extreme_event()` except that the variables defined in the `set_values_to` field of `event()` or `event_joined()` are added to the input dataset.

`derive_var_dthcause()`, `derive_var_extreme_dt()`, and `derive_var_extreme_dtm()` should be superseded.

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by jeffreyad**
Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by jeffreyad** [comment]
@bundfussr I'm not sure if this is exactly what you had in mind, but it seemed like the best option was to create a new helper function `derive_extreme()` which does most of what `derive_extreme_event()` previously did and then call that function in two new wrapper functions `derive_extreme_event()` and `derive_vars_extreme_event()` to either add the new records as new observations or new variables.

**Comment by bundfussr** [comment]
> @bundfussr I'm not sure if this is exactly what you had in mind, but it seemed like the best option was to create a new helper function `derive_extreme()` which does most of what `derive_extreme_event()` previously did and then call that function in two new wrapper functions `derive_extreme_event()` and `derive_vars_extreme_event()` to either add the new records as new observations or new variables.

Sounds good to me. I'll have a closer look later today or on Monday.

File: NEWS.md [hunk 1]
```diff
@@ -2,11 +2,10 @@
 
 ## New Features
 
-New global option created `signif_digits` to handle floating point issue, the value is
-set to `15`, and is used with the `base R` function `signif()` when comparing 2 numeric
-values. This is implemented in `admiral ` functions `derive_var_atoxgr_dir()` and `derive_var_anrind()`. (#2134)
+- Implement `derive_vars_extreme_event()`, which works as `derive_extreme_event()` but returns variables instead of a parameter. (#2138)
```
Comment (bms63) [inline NEWS.md:5]: Can we make a mention of this superseding functions as well?

File: R/derive_extreme_event.R
Comment (bms63) [inline R/derive_extreme_event.R:1]: What do you think about splitting out these functions into different files?  suggestion event, event_joined gets one file, derive_extreme gets one file and the two main functions get separate files.

This file is up to 1460 lines!! whew!  :)

Comment (bms63) [reply R/derive_extreme_event.R:1]: The tests files would also get updated to correspond to the new naming scheme

File: tests/testthat/test-derive_extreme_event.R
Comment (bms63) [inline tests/testthat/test-derive_extreme_event.R:1]: Is there any chance (if you have time) we could get this line covered? 
image

File: NEWS.md (hunk 1 above)
Comment (bms63) [reply NEWS.md:5]: and list out the functions it is superseding

**Comment by bms63** [comment]
Do we want users to have access to this help function?  I'm thinking no??? 

image

File: tests/testthat/test-derive_extreme_event.R
Comment (jeffreyad) [reply tests/testthat/test-derive_extreme_event.R:1]: Added test for this.

**Comment by bms63** [comment]
I think if you add pharmaversesdtm to the Remotes you can get the new changes there so the R-CMD checks pass again?  @ddsjoberg correct?

image

**Comment by ddsjoberg** [comment]
@bms63 which R CMD Check error are you trying to solve with the `Remotes:`?

File: R/derive_extreme_event.R [hunk 2]
```diff
@@ -482,9 +485,591 @@ derive_extreme_event <- function(dataset,
...
+derive_vars_extreme_event <- function(dataset,
+                                      by_vars = NULL,
+                                      events,
+                                      tmp_event_nr_var = NULL,
+                                      order,
+                                      mode,
+                                      source_datasets = NULL,
+                                      ignore_event_order = NULL,
+                                      check_type = "warning",
+                                      set_values_to,
```
Comment (bundfussr) [inline R/derive_extreme_event.R:1052]: Should be changed to `new_vars`.

File: R/derive_var_dthcaus.R [hunk 3]
```diff
@@ -1,5 +1,9 @@
 #' Derive Death Cause
 #'
+#' @description
+#' `r lifecycle::badge("superseded")` The `derive_var_dthcaus()`
+#' function has been superseded in favor of `derive_vars_extreme_event()`.
```
Comment (bundfussr) [inline R/derive_var_dthcaus.R:5]: Please update `@family` and `@keyword` to `superseded`.

File: R/derive_var_extreme_date.R [hunk 4]
```diff
@@ -1,5 +1,9 @@
 #' Derive First or Last Datetime from Multiple Sources
 #'
+#' @description
+#' `r lifecycle::badge("superseded")` The `derive_var_extreme_dtm()`
+#' function has been superseded in favor of `derive_vars_extreme_event()`.
```
Comment (bundfussr) [inline R/derive_var_extreme_date.R:5]: Please update `@family` and `@keyword` to `superseded`.

File: R/derive_extreme_event.R [hunk 5]
```diff
@@ -482,9 +485,591 @@ derive_extreme_event <- function(dataset,
...
+#'       set_values_to = exprs(LSTALVDT = TRTEDT, DTHFL = "N")
+#'     )
+#'   ),
+#'   source_datasets = list(adsl = adsl, lb = lb),
+#'   tmp_event_nr_var = event_nr,
+#'   order = exprs(USUBJID, LSTALVDT, tmp_event_nr_var),
+#'   mode = "last",
+#'   keep_source_vars = exprs(STUDYID, USUBJID),
+#'   set_values_to = exprs(LSTALVDT = LSTALVDT, DTHFL = DTHFL)
+#' )
```
Comment (bundfussr) [inline R/derive_extreme_event.R:1042]: Please add an example for `DTHCAUS`.

File: R/derive_extreme_event.R [hunk 6]
```diff
@@ -482,9 +485,591 @@ derive_extreme_event <- function(dataset,
...
+#'       dataset_name = "adsl",
+#'       condition = !is.na(TRTEDT),
+#'       order = exprs(TRTEDT),
+#'       mode = "last",
+#'       set_values_to = exprs(LSTALVDT = TRTEDT, DTHFL = "N")
+#'     )
+#'   ),
+#'   source_datasets = list(adsl = adsl, lb = lb),
+#'   tmp_event_nr_var = event_nr,
+#'   order = exprs(USUBJID, LSTALVDT, tmp_event_nr_var),
```
Comment (bundfussr) [inline R/derive_extreme_event.R:1038]: Should be `event_nr` instead of `tmp_event_nr_var`.

**Review by bundfussr — CHANGES_REQUESTED** [review]
- Please update the ADSL vignette and template.
- Please supersede `dthcaus_source` and `date_source`.

**Comment by bms63** [comment]
> @bms63 which R CMD Check error are you trying to solve with the `Remotes:`?

My understanding is this is related to a recent update in pharmaversesdtm

image

**Comment by bundfussr** [comment]
> > @bms63 which R CMD Check error are you trying to solve with the `Remotes:`?
> 
> My understanding is this is related to a recent update in pharmaversesdtm
> 
> image

Unfortunately, the development version is broken at the moment. We need to merge https://github.com/pharmaverse/pharmaversesdtm/pull/78 and https://github.com/pharmaverse/admiral/pull/2241 to fix it.

File: R/derive_extreme_event.R [hunk 7]
```diff
@@ -340,17 +343,17 @@
...
-                                 events,
-                                 tmp_event_nr_var = NULL,
-                                 order,
-                                 mode,
-                                 source_datasets = NULL,
-                                 ignore_event_order = FALSE,
-                                 check_type = "warning",
-                                 set_values_to,
-                                 keep_source_vars = exprs(everything())) {
+derive_extreme <- function(dataset,
```
[the remark this answers is not in the document]
Comment (jeffreyad) [reply R/derive_extreme_event.R]: @bundfussr Thanks, I removed `derive_extreme()` and call `derive_extreme_event()` in `derive_vars_extreme_event()`.

File: R/derive_extreme_event.R
Comment (jeffreyad) [reply R/derive_extreme_event.R:1]: Removed `derive_extreme()` so it is smaller now.

File: R/derive_extreme_event.R [hunk 8]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+                                      check_type = "warning",
+                                      new_vars,
+                                      keep_source_vars = by_vars) {
+  assert_expr_list(new_vars)
+
+  new_obs <- derive_extreme_event(
+    dataset = NULL,
+    by_vars = by_vars,
+    events = events,
+    tmp_event_nr_var = event_nr,
```
Comment (bundfussr) [inline R/derive_extreme_event.R:712]: Should be `!!tmp_event_nr_var` instead of `event_nr`.

File: R/derive_extreme_event.R [hunk 9]
```diff
@@ -347,18 +347,15 @@ derive_extreme_event <- function(dataset,
                                  order,
                                  mode,
                                  source_datasets = NULL,
-                                 ignore_event_order = FALSE,
+                                 ignore_event_order = NULL,
                                  check_type = "warning",
                                  set_values_to,
                                  keep_source_vars = exprs(everything())) {
   # Check input parameters
+  assert_data_frame(dataset, optional = TRUE)
```
Comment (bundfussr) [inline R/derive_extreme_event.R:355]: The default value of `dataset` (in line 343) should be set to `NULL`.

File: R/derive_extreme_event.R [hunk 10]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+  new_obs <- derive_extreme_event(
+    dataset = NULL,
+    by_vars = by_vars,
+    events = events,
+    tmp_event_nr_var = event_nr,
+    order = order,
+    mode = mode,
+    source_datasets = source_datasets,
+    check_type = check_type,
+    set_values_to = new_vars,
```
Comment (bundfussr) [inline R/derive_extreme_event.R:717]: I would remove this line and pass `new_vars` to the `new_vars` argument in the `derive_vars_merged()` call below.

File: R/derive_extreme_event.R [hunk 11]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+derive_vars_extreme_event <- function(dataset,
+                                      by_vars = NULL,
+                                      events,
+                                      tmp_event_nr_var = NULL,
+                                      order,
+                                      mode,
+                                      source_datasets = NULL,
+                                      check_type = "warning",
+                                      new_vars,
+                                      keep_source_vars = by_vars) {
```
Comment (bundfussr) [inline R/derive_extreme_event.R:705]: The default value does not make sense as the by variables are kept anyway. I would remove the `keep_source_vars` argument. I think it is clearer if the new variables are specified by `new_vars` only.

File: R/derive_extreme_event.R [hunk 12]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+#'     event(
+#'       dataset_name = "ds",
+#'       condition = DSDECOD == "DEATH" & grepl("DEATH DUE TO", DSTERM),
+#'       set_values_to = exprs(DTHCAUS = DSTERM, DTHDT = convert_dtc_to_dt(DSSTDTC)),
+#'       order = exprs(DTHDT)
+#'     )
+#'   ),
+#'   source_datasets = list(ae = ae, ds = ds),
+#'   tmp_event_nr_var = event_nr,
+#'   order = exprs(USUBJID, DTHDT, event_nr),
```
Comment (bundfussr) [inline R/derive_extreme_event.R:691]: No need to specify `USUBJID` because it is already in `by_vars`.

File: R/derive_extreme_event.R [hunk 13]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+#'     event(
+#'       dataset_name = "adsl",
+#'       condition = !is.na(TRTEDT),
+#'       order = exprs(TRTEDT),
+#'       mode = "last",
+#'       set_values_to = exprs(LSTALVDT = TRTEDT, DTHFL = "N")
+#'     )
+#'   ),
+#'   source_datasets = list(adsl = adsl, lb = lb),
+#'   order = exprs(USUBJID, LSTALVDT, event_nr),
```
Comment (bundfussr) [inline R/derive_extreme_event.R:645]: No need to specify `USUBJID` because it is already in `by_vars`.

File: R/derive_extreme_event.R [hunk 14]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+#'   mode = "last",
+#'   keep_source_vars = exprs(STUDYID, USUBJID),
```
Comment (bundfussr) [inline R/derive_extreme_event.R:647]: Should be removed.

File: R/derive_extreme_event.R [hunk 15]
```diff
@@ -486,6 +482,248 @@ derive_extreme_event <- function(dataset,
...
+#'       condition = DSDECOD == "DEATH" & grepl("DEATH DUE TO", DSTERM),
+#'       set_values_to = exprs(DTHCAUS = DSTERM, DTHDT = convert_dtc_to_dt(DSSTDTC)),
+#'       order = exprs(DTHDT)
+#'     )
+#'   ),
+#'   source_datasets = list(ae = ae, ds = ds),
+#'   tmp_event_nr_var = event_nr,
+#'   order = exprs(USUBJID, DTHDT, event_nr),
+#'   mode = "first",
+#'   keep_source_vars = exprs(STUDYID, USUBJID),
```
Comment (bundfussr) [inline R/derive_extreme_event.R:693]: Should be removed.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_extreme_event.R (hunk 8 above)
Comment (jeffreyad) [reply R/derive_extreme_event.R]: Updated, thanks!

File: R/derive_extreme_event.R (hunk 9 above)
Comment (jeffreyad) [reply R/derive_extreme_event.R:355]: Updated, thanks!

File: R/derive_extreme_event.R (hunk 10 above)
Comment (jeffreyad) [reply R/derive_extreme_event.R]: I tried removing `set_values_to` but that seemed to result in an error if it wasn't assigned.

File: R/derive_extreme_event.R (hunk 11 above)
Comment (jeffreyad) [reply R/derive_extreme_event.R]: Updated, thanks!

File: R/derive_extreme_event.R (hunk 12 above)
Comment (jeffreyad) [reply R/derive_extreme_event.R]: Updated

**Comment by manciniedoardo** [comment]
Hi all - hope you don't mind me removing myself from review of this as I won't have time to do a thorough review - please feel free to ask someone else if required.

File: R/derive_extreme_event.R (hunk 10 above)
Comment (bundfussr) [reply R/derive_extreme_event.R]: I see. I would suggest changing the `set_value_to` argument in `derive_extreme_event()` to optional.

File: inst/templates/ad_adsl.R [hunk 16]
```diff
@@ -194,29 +194,49 @@ adsl <- adsl %>%
...
+        condition = !is.na(TRTEDT),
+        set_values_to = exprs(LSTALVDT = TRTEDT, seq = NA_integer_),
+      )
+    ),
     source_datasets = list(ae = ae, lb = lb, adsl = adsl),
-    mode = "last"
+    tmp_event_nr_var = event_nr,
+    order = exprs(LSTALVDT, seq, event_nr),
+    mode = "last",
+    new_vars = exprs(LSTALVDT = LSTALVDT)
```
Comment (bundfussr) [inline inst/templates/ad_adsl.R:239]: Can be simplified to `new_vars = exprs(LSTALVDT)`.

File: R/derive_extreme_event.R [hunk 17]
```diff
@@ -486,6 +482,234 @@ derive_extreme_event <- function(dataset,
...
+#'       order = exprs(TRTEDT),
+#'       mode = "last",
+#'       set_values_to = exprs(LSTALVDT = TRTEDT, DTHFL = "N")
+#'     )
+#'   ),
+#'   source_datasets = list(adsl = adsl, lb = lb),
+#'   tmp_event_nr_var = event_nr,
+#'   order = exprs(LSTALVDT, event_nr),
+#'   mode = "last",
+#'   new_vars = exprs(LSTALVDT = LSTALVDT, DTHFL = DTHFL)
```
Comment (bundfussr) [inline R/derive_extreme_event.R:635]: Can be simplified to `new_vars = exprs(LSTALVDT, DTHFL)`.

File: vignettes/adsl.Rmd [hunk 18]
```diff
@@ -415,72 +415,40 @@ See also [Date and Time Imputation](imputation.html).
...
+        dataset_name = "ds",
+        condition = DSDECOD == "DEATH" & grepl("DEATH DUE TO", DSTERM),
+        set_values_to = exprs(DTHCAUS = DSTERM),
+      )
+    ),
+    source_datasets = list(ae = ae, ds = ds),
+    tmp_event_nr_var = event_nr,
+    order = exprs(event_nr),
+    mode = "first",
+    new_vars = exprs(DTHCAUS = DTHCAUS)
```
Comment (bundfussr) [inline vignettes/adsl.Rmd:451]: Can be simplified to `new_vars = exprs(DTHCAUS)`.

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_extreme_event.R (hunk 17 above)
Comment (jeffreyad) [reply R/derive_extreme_event.R:635]: Updated, thanks!

File: inst/templates/ad_adsl.R (hunk 16 above)
Comment (jeffreyad) [reply inst/templates/ad_adsl.R]: Updated, thanks!

File: R/derive_extreme_event.R [hunk 19]
```diff
@@ -486,6 +482,234 @@ derive_extreme_event <- function(dataset,
...
+#'       order = exprs(DTHDT)
+#'     )
+#'   ),
+#'   source_datasets = list(ae = ae, ds = ds),
+#'   tmp_event_nr_var = event_nr,
+#'   order = exprs(DTHDT, event_nr),
+#'   mode = "first",
+#'   new_vars = exprs(DTHCAUS = DTHCAUS, DTHDT = DTHDT)
+#' )
+derive_vars_extreme_event <- function(dataset,
```
[the remark this answers is not in the document]
Comment (jeffreyad) [reply R/derive_extreme_event.R]: @bms63 Ok, I split into three files.

**Comment by jeffreyad** [comment]
> * Please update the ADSL vignette and template.
> * Please supersede `dthcaus_source` and `date_source`.

Superseded `dthcaus_source` and `date_source`.

File: R/derive_extreme_event.R [hunk 20]
```diff
@@ -340,25 +340,22 @@
...
                                  by_vars = NULL,
                                  events,
                                  tmp_event_nr_var = NULL,
                                  order,
                                  mode,
                                  source_datasets = NULL,
-                                 ignore_event_order = FALSE,
+                                 ignore_event_order = NULL,
                                  check_type = "warning",
                                  set_values_to,
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_extreme_event.R:352]: @jeffreyad , I can't see the change. It seems there is still no default value for `set_values_to`.

[the remark this answers is not in the document]
Comment (jeffreyad) [reply R/derive_extreme_event.R]: Got it, updated.

File: R/derive_vars_extreme_event.R [hunk 21]
```diff
@@ -0,0 +1,228 @@
...
+#' @param events Conditions and new values defining events
+#'
+#'   A list of `event()` or `event_joined()` objects is expected. Only
+#'   observations listed in the `events` are considered for deriving extreme
+#'   event. If multiple records meet the filter `condition`, take the first
+#'   record sorted by `order`. The data is grouped by `by_vars`, i.e., summary
+#'   functions like `all()` or `any()` can be used in `condition`.
+#'
+#'   For `event_joined()` events the observations are selected by calling
+#'   `filter_joined()`. The `condition` field is passed to the `filter` argument.
```
Comment (bundfussr) [inline R/derive_vars_extreme_event.R:16]: Should be `filter_join` instead of `filter` as `filter` is deprecated.

**Review by bundfussr — CHANGES_REQUESTED** [review]
Almost done. Just one comment left.

File: R/derive_vars_extreme_event.R (hunk 21 above)
Comment (jeffreyad) [reply R/derive_vars_extreme_event.R]: Updated

**Review by bundfussr — APPROVED** [review]
Thanks for your patience.

**Comment by jeffreyad** [comment]
> Thanks for your patience.

Thanks!  I thought of one more thing: I could remove `set_values_to` from `derive_vars_extreme_event()` since we set it to NULL in `derive_extreme_event()`.

**Review by bms63 — APPROVED** [review]
I'm happy!  @jeffreyad @bundfussr shall we click the button!!??!!

**Pull request merged**